ocaml/xapi/storage_impl.ml

```diff
@@ -619,24 +622,80 @@ module Wrapper = functor(Impl: Server_impl) -> struct
         		    the resources associated with [dp] in [sr]. If [vdi_already_locked] then
         		    it is assumed that all VDIs are already locked. *)
     let destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak vdi_already_locked =
-      (* Every VDI in use by this session should be detached and deactivated *)
+      (* Every VDI in use by this session should be detached and deactivated
+	 This code makes the assumption that a datapath is only on 0 or 1 VDIs. However, it retains debug code (identified below) to verify this.
+         It also assumes that the VDIs associated with a datapath don't change during its execution - again it retains debug code to verify this.	
+      *)
+
       let vdis = Sr.list sr_t in
-      List.fold_left (fun acc (vdi, vdi_t) ->
-          let locker =
-            if vdi_already_locked
-            then fun f -> f ()
-            else VDI.with_vdi sr vdi in
-          locker
-            (fun () ->
-               try
-                 VDI.destroy_datapath_nolock context ~dbg ~dp ~sr ~vdi ~allow_leak;								acc
-               with e -> e::acc
-            )) [] vdis
+      
+      (* Note that we assume this filter returns 0 or 1 items, but we need to verify that. *)
+      let vdis_with_dp = List.filter (fun(vdi, vdi_t) -> Vdi.dp_on_vdi dp vdi_t) vdis in
+      debug "[destroy_sr] Filtered VDI count:%d" (List.length vdis_with_dp);
+      List.iter (fun(vdi, vdi_t) -> debug "[destroy_sr] VDI found with the dp is %s" vdi) vdis_with_dp;
+      
+      let locker vdi =
+        if vdi_already_locked
+          then fun f -> f ()
+          else VDI.with_vdi sr vdi in
+
+      (* This is debug code to verify that no more than 1 VDI matched the datapath. We also convert the 0 and 1 cases to an Option which is more natural to work with *)
+      let vdi_to_remove = match vdis_with_dp with
+        | [] -> None
+        | [x] -> Some x
+        | _ -> 
+           raise (Storage_interface.Backend_error (Api_errors.internal_error, [Printf.sprintf "Expected 0 or 1 VDI with datapath, had %d" 1]));
```

Comment:
Looks like a count to the number of VDIs to display in the %d is missing and there is a 1 instead